src/config/graph.rs

```diff
@@ -139,28 +141,40 @@ impl Graph {
         }
     }
 
+    fn get_input_type(&self, key: &ComponentKey) -> DataType {
+        match self.nodes[key] {
+            Node::Source { .. } => panic!("no inputs on sources"),
+            Node::Transform { in_ty, .. } => in_ty,
+            Node::Sink { ty } => ty,
+        }
+    }
+
+    fn get_output_type(&self, id: &OutputId) -> DataType {
+        match &self.nodes[&id.component] {
+            Node::Source { outputs } | Node::Transform { outputs, .. } => outputs
+                .iter()
+                .find(|output| output.port == id.port)
+                .map(|output| output.ty)
+                .expect("output didn't exist"),
+            Node::Sink { .. } => panic!("no outputs on sinks"),
+        }
+    }
+
     pub fn typecheck(&self) -> Result<(), Vec<String>> {
         let mut errors = Vec::new();
 
         // check that all edges connect components with compatible data types
         for edge in &self.edges {
-            let (in_key, out_key) = (&edge.from.component, &edge.to);
-            if self.nodes.get(in_key).is_none() || self.nodes.get(out_key).is_none() {
-                continue;
-            }
-            match (self.nodes[in_key].clone(), self.nodes[out_key].clone()) {
-                (Node::Source { ty: ty1 }, Node::Sink { ty: ty2, .. })
-                | (Node::Source { ty: ty1 }, Node::Transform { in_ty: ty2, .. })
-                | (Node::Transform { out_ty: ty1, .. }, Node::Transform { in_ty: ty2, .. })
-                | (Node::Transform { out_ty: ty1, .. }, Node::Sink { ty: ty2, .. }) => {
-                    if ty1 != ty2 && ty1 != DataType::Any && ty2 != DataType::Any {
-                        errors.push(format!(
-                            "Data type mismatch between {} ({:?}) and {} ({:?})",
-                            in_key, ty1, out_key, ty2
-                        ));
-                    }
-                }
-                (Node::Sink { .. }, _) | (_, Node::Source { .. }) => unreachable!(),
+            // TODO: these panic on unknown nodes, where previously we'd skip them. which is
+            // correct?
```

Comment:
So long as we have a good error message going out to the user my general preference for unrecoverable bugs -- which I think this counts as -- is to crash, yeah. So I'm for this.

Reply:
Ah yeah, I'd mean to trace this back and double check what our assumptions should be when this gets called. Looks like it's only called as one of the final steps of compiling a config, so we'll have already checked that edges are valid. Seems like a panic is appropriate, so I'll swap out the TODO for a doc comment.